internal/grpc/commander/connection.go

```diff
@@ -0,0 +1,203 @@
+	cmdExchanger exchanger.CommandExchanger
+	logger       logr.Logger
+	id           string
+	nginxID      string
+	systemID     string
+	state        State
+}
+
+func (c *connection) ID() string {
+	return c.id
+}
+
+func (c *connection) State() State {
+	return c.state
+}
+
+// ReceiveFromUploadServer uploads data chunks from the UploadServer and logs them.
+// FIXME(kate-osborn): NKG doesn't need this functionality and ideally we wouldn't have to implement and maintain this.
+// Figure out how to remove this without causing errors in the agent.
+func (c *connection) ReceiveFromUploadServer(server proto.Commander_UploadServer) error {
+	c.logger.Info("Upload request")
+
+	for {
+		_, err := server.Recv()
+
+		if err != nil && !errors.Is(err, io.EOF) {
+			c.logger.Error(err, "upload receive error")
+			return err
+		}
+
+		c.logger.Info("Received chunk from upload channel")
+
+		if errors.Is(err, io.EOF) {
+			c.logger.Info("Upload completed")
+			return server.SendAndClose(&proto.UploadStatus{Status: proto.UploadStatus_OK})
+		}
+	}
+}
+
+// newConnection creates a new instance of connection.
+//
+// id is the unique ID of the connection.
+// cmdExchanger sends and receives commands to and from the CommandChannelServer.
+//
+// The creator of connection must call its run method in order for the connection send and receive commands.
+func newConnection(
+	id string,
+	logger logr.Logger,
+	cmdExchanger exchanger.CommandExchanger,
+) *connection {
+	return &connection{
+		logger:       logger.WithName(fmt.Sprintf("connection-%s", id)),
+		cmdExchanger: cmdExchanger,
+		id:           id,
+	}
+}
+
+// run is a blocking method that kicks off the connection's receive loop and the CommandExchanger's Run loop.
+// run will return when the context is canceled or if either loop returns an error.
+func (c *connection) run(parent context.Context) error {
+	eg, ctx := errgroup.WithContext(parent)
+
+	eg.Go(func() error {
+		return c.receive(ctx)
+	})
+
+	eg.Go(func() error {
+		return c.cmdExchanger.Run(ctx)
+	})
+
+	return eg.Wait()
+}
+
+func (c *connection) receive(ctx context.Context) error {
+	defer func() {
+		c.logger.Info("Stopping receive loop")
+	}()
+	c.logger.Info("Starting receive loop")
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case cmd := <-c.cmdExchanger.Out():
+			c.handleCommand(cmd)
+		}
+	}
+}
+
+func (c *connection) handleCommand(cmd *proto.Command) {
+	if cmd == nil {
+		return
+	}
+
+	switch cmd.Data.(type) {
+	case *proto.Command_AgentConnectRequest:
+		c.handleAgentConnectRequest(cmd)
+	default:
+		c.logger.Info("Ignoring command", "command data type", fmt.Sprintf("%T", cmd.Data))
+	}
+}
+
+func (c *connection) handleAgentConnectRequest(cmd *proto.Command) {
+	req := cmd.GetAgentConnectRequest()
+
+	c.logger.Info("Received agent connect request", "message ID", cmd.GetMeta().GetMessageId())
+
+	c.register(getFirstNginxID(req.GetDetails()), req.GetMeta().SystemUid)
+
+	res := createAgentConnectResponseCmd(
```

Comment:
That's a good question. It shouldn't be possible for either field to be empty.  This check is a relic from when I attempted to re-register the agent using the `DataplaneStatus` command. 

I can remove the check if we feel it's safe to trust that the agent will always send these IDs in the registration payload. 

Otherwise, I think we have the following options:
(1) Panic. It's not our programming error, but it's the agents, which is part of our product.
(2) Don't send a response. As far as I can tell, this won't negatively affect us. 
(3) Send the response but set the StatusCode to Rejected. The agent doesn't check the StatusCode, so it won't change any behavior, but it will add context to the logs of both components, which would be helpful with debugging. 

What do you think?

Reply:
Spoke offline. We are going to go with option #3.

Reply:
Implemented option #3. I also removed the `AgentConfig` from the response since we don't have any use for it, which removed the need to implement the `Upload` rpc. The agent only uploads its config on registration if the `AgentConfig` is provided. Otherwise, `Upload` has to be triggered by the control plane. If we find out later that we need the `AgentConfig`, I can add the `Upload` functionality back.